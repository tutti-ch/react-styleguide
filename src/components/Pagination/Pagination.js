import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  ArrowStart,
  ArrowEnd,
  Next,
  Prev
} from "../../styles/Icons/assets/generic";
import classes from "./Pagination.scss";
import Page from "./_Page";

export default class Pagination extends Component {
  static propTypes = {
    /**
     * Indicates the current active page
     */
    page: PropTypes.number.isRequired,

    /**
     * Number of total items that needs to be paginated
     */
    totalItems: PropTypes.number,

    /**
     * Text shown before the page number (only mobile view)
     */
    pageText: PropTypes.string,

    /**
     * Function called when a page number is clicked
     */
    onChange: PropTypes.func,

    /**
     * Number of items shown on every page
     */
    itemsPerPage: PropTypes.number,

    /**
     * Number of pages visible at the same time to the user
     */
    pageBucket: PropTypes.number,

    // Generates the URL for the single pages
    getPageUrl: PropTypes.func
  };

  static defaultProps = {
    itemsPerPage: 10,
    pageBucket: 10,
    onChange: () => {},
    getPageUrl: () => "#"
  };

  render() {
    const {
      totalItems,
      itemsPerPage,
      page,
      pageBucket,
      onChange,
      pageText,
      getPageUrl
    } = this.props;
    const pagesCount = Math.ceil(totalItems / itemsPerPage);
    const currentBucket = Math.floor((page - 0.5) / pageBucket);

    const startPage = Math.max(currentBucket * pageBucket, 1);
    const endingPage = Math.min(
      currentBucket * pageBucket + pageBucket,
      pagesCount
    );

    const pages = [];
    for (let o = startPage; o <= endingPage; o++) {
      pages.push(
        <Page
          key={`pagination-${o}`}
          number={o}
          active={page === o}
          onClick={onChange}
          url={getPageUrl(o)}
        />
      );
    }
    const nextPage = page + 1;
    const isTooBig = nextPage > pagesCount;
    const isLast = page === pagesCount;

    const previousPage = page - 1;
    const isTooSmall = previousPage < 1;
    const isFirst = page === 1;

    return (
      <>
        {pagesCount > 1 && (
          <nav role="navigation" aria-label="Pagination Navigation">
            <ul className={classes.pagination_bar}>
              <Page
                number={1}
                disabled={isFirst}
                onClick={onChange}
                icon={ArrowStart}
                url={getPageUrl(1)}
              />
              <Page
                number={isTooSmall ? 1 : previousPage}
                disabled={isTooSmall}
                icon={Prev}
                onClick={onChange}
                url={getPageUrl(isTooSmall ? 1 : previousPage)}
              />
              <li className={classes.text}>{pageText}</li>
              {pages}
              <Page
                number={isTooBig ? pagesCount : nextPage}
                disabled={isTooBig}
                icon={Next}
                onClick={onChange}
                url={getPageUrl(isTooBig ? pagesCount : nextPage)}
              />
              <Page
                number={pagesCount}
                disabled={isLast}
                icon={ArrowEnd}
                onClick={onChange}
                url={getPageUrl(pagesCount)}
              />
            </ul>
          </nav>
        )}
      </>
    );
  }
}
