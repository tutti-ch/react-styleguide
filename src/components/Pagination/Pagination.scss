@import "main";

.pagination_bar {
  width: $phone;
  margin: $size-m auto $size-xl auto;
  display: flex;
  justify-content: center;
  //   @include clearfix;

  @include tablet-and-bigger {
    width: 100%;
    margin: $size-m 0 $size-xl 0;
  }

  @include phone-and-smaller {
    width: 100%; // To avoid breaking layout in small phones
  }
}

.page {
  display: block;
  text-align: center;
  height: 100%;
  width: 20%; // Can't use flex: 1 (with max-width) because breaks in ie11

  @include tablet-and-bigger {
    width: 7.14%; // 100% divided in 14 blocks
    // Can't use flexbox with max-width because breaks in ie11
  }

  &:hover {
    cursor: pointer;
  }
}

.active {
  width: auto;
  padding-left: $size-xxs;
  color: $greyscale-20;

  @include tablet-and-bigger {
    width: 7.14%; // 100% divided in 14 blocks
    padding-left: 0;
    background: $greyscale-95;
  }
}

.inactive {
  display: none;

  @include tablet-and-bigger {
    display: block;
  }
}

.disabled {
  opacity: 0.3;

  .link:hover {
    cursor: default;
  }
}

.active,
.disabled,
body.offline .page {
  pointer-events: none;
}

.link {
  width: 100%;
  height: 100%;
  display: block;
  line-height: $size-xxl !important;
}

// the text 'page' appears for small screens
.text {
  float: left;
  display: block;
  color: $blue-40;
  text-align: center;
  line-height: $size-xxl;

  @include tablet-and-bigger {
    display: none;
  }
}
