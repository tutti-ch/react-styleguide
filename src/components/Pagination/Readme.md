```jsx
<Pagination page={1} totalItems={3020} pageText='Page' />
```

```jsx
<Pagination page={145} totalItems={3020} pageText='Page' />
```

```jsx
<Pagination page={302} totalItems={3020} pageText='Page' />
```