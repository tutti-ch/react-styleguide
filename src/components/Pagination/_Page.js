import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import classes from "./Pagination.scss";

export default class Page extends Component {
  static propTypes = {
    number: PropTypes.number.isRequired,
    icon: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
    disabled: PropTypes.bool,
    active: PropTypes.bool,
    onClick: PropTypes.func,
    url: PropTypes.string
  };

  static defaultProps = {
    url: "#"
  };

  handleClick = number => e => {
    e.preventDefault();
    const { onClick } = this.props;

    if (typeof onClick === "function") {
      this.props.onClick(number);
    }
  };

  render() {
    const { number, icon: Icon, active, disabled, url } = this.props;

    const listItemClasses = classNames(classes.page, {
      [classes.active]: active,
      [classes.inactive]: active === false,
      [classes.disabled]: disabled
    });

    const linkClasses = classNames("link", [classes.link]);

    return (
      <li className={listItemClasses}>
        <a
          href={url}
          className={linkClasses}
          aria-label={`Go to page ${number}`}
          onClick={this.handleClick(number)}
        >
          {Icon ? <Icon /> : number}
        </a>
      </li>
    );
  }
}
