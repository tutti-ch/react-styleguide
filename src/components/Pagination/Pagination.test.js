/* global describe, test */
import React from "react";
import { shallow, mount } from "enzyme";
import Pagination from "./index";
import toJson from "enzyme-to-json";

describe("(Component) Pagination", () => {
  test("should match the snapshot", () => {
    const comp = mount(
      <Spinner
        size={Spinner.SIZE_LARGE}
        color={Spinner.COLOR_DARK}
        className="my-class"
      />
    );
    expect(toJson(comp)).toMatchSnapshot();
  });
});
