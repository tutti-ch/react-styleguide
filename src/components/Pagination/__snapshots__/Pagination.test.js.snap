// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Pagination should match the snapshot 1`] = `
<Pagination
  getPageUrl={[Function]}
  itemsPerPage={10}
  onChange={[Function]}
  page={302}
  pageBucket={10}
  pageText="Page"
  totalItems={3020}
>
  <nav
    aria-label="Pagination Navigation"
    role="navigation"
  >
    <ul>
      <Page
        disabled={false}
        icon={[Function]}
        number={1}
        onClick={[Function]}
        url="#"
      >
        <li
          className=""
        >
          <a
            aria-label="Go to page 1"
            className="link"
            href="#"
            onClick={[Function]}
          >
            <ArrowStart>
              <Sprite
                id="arrow-start"
                viewBox="0 0 15.4 30"
              >
                <svg
                  className="svg-sprite"
                  role="img"
                  viewBox="0 0 15.4 30"
                >
                  <use
                    xlinkHref="#arrow-start"
                  />
                </svg>
              </Sprite>
            </ArrowStart>
          </a>
        </li>
      </Page>
      <Page
        disabled={false}
        icon={[Function]}
        number={301}
        onClick={[Function]}
        url="#"
      >
        <li
          className=""
        >
          <a
            aria-label="Go to page 301"
            className="link"
            href="#"
            onClick={[Function]}
          >
            <Prev>
              <Sprite
                id="prev"
                viewBox="0 0 30 30"
              >
                <svg
                  className="svg-sprite"
                  role="img"
                  viewBox="0 0 30 30"
                >
                  <use
                    xlinkHref="#prev"
                  />
                </svg>
              </Sprite>
            </Prev>
          </a>
        </li>
      </Page>
      <li>
        Page
      </li>
      <Page
        active={false}
        key="pagination-300"
        number={300}
        onClick={[Function]}
        url="#"
      >
        <li
          className=""
        >
          <a
            aria-label="Go to page 300"
            className="link"
            href="#"
            onClick={[Function]}
          >
            300
          </a>
        </li>
      </Page>
      <Page
        active={false}
        key="pagination-301"
        number={301}
        onClick={[Function]}
        url="#"
      >
        <li
          className=""
        >
          <a
            aria-label="Go to page 301"
            className="link"
            href="#"
            onClick={[Function]}
          >
            301
          </a>
        </li>
      </Page>
      <Page
        active={true}
        key="pagination-302"
        number={302}
        onClick={[Function]}
        url="#"
      >
        <li
          className=""
        >
          <a
            aria-label="Go to page 302"
            className="link"
            href="#"
            onClick={[Function]}
          >
            302
          </a>
        </li>
      </Page>
      <Page
        disabled={true}
        icon={[Function]}
        number={302}
        onClick={[Function]}
        url="#"
      >
        <li
          className="undefined"
        >
          <a
            aria-label="Go to page 302"
            className="link"
            href="#"
            onClick={[Function]}
          >
            <Next>
              <Sprite
                id="next"
                viewBox="0 0 30 30"
              >
                <svg
                  className="svg-sprite"
                  role="img"
                  viewBox="0 0 30 30"
                >
                  <use
                    xlinkHref="#next"
                  />
                </svg>
              </Sprite>
            </Next>
          </a>
        </li>
      </Page>
      <Page
        disabled={true}
        icon={[Function]}
        number={302}
        onClick={[Function]}
        url="#"
      >
        <li
          className="undefined"
        >
          <a
            aria-label="Go to page 302"
            className="link"
            href="#"
            onClick={[Function]}
          >
            <ArrowEnd>
              <Sprite
                id="arrow-end"
                viewBox="0 0 15.4 30"
              >
                <svg
                  className="svg-sprite"
                  role="img"
                  viewBox="0 0 15.4 30"
                >
                  <use
                    xlinkHref="#arrow-end"
                  />
                </svg>
              </Sprite>
            </ArrowEnd>
          </a>
        </li>
      </Page>
    </ul>
  </nav>
</Pagination>
`;
